src/ore/src/region.rs

```diff
@@ -148,3 +150,232 @@ impl<T> LgAllocRegion<T> {
         }
     }
 }
+
+/// An abstraction over different kinds of allocated regions.
+///
+/// # WARNING
+///
+/// The implementation does not drop its elements, but forgets them instead. Do not use where
+/// this is not intended, i.e., outside `Copy` types or columnation regions.
+#[derive(Debug)]
+pub enum Region<T> {
+    /// A possibly empty heap-allocated region, represented as a vector.
+    Heap(Vec<T>),
+    /// A mmaped region, represented by a vector and its backing memory mapping.
+    MMap(MMapRegion<T>),
+}
+
+/// Type encapsulating private data for memory-mapped regions.
+pub struct MMapRegion<T> {
+    /// Vector-representation of the underlying memory. Must not be dropped.
+    inner: ManuallyDrop<Vec<T>>,
+    /// Opaque handle to lgalloc.
+    handle: Option<lgalloc::Handle>,
+}
+
+impl<T> MMapRegion<T> {
+    /// Clear the contents of this region without dropping elements.
+    unsafe fn clear(&mut self) {
+        self.inner.set_len(0);
+    }
+}
+
+impl<T: Debug> Debug for MMapRegion<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("MMapRegion")
+            .field("inner", &self.inner)
+            .finish_non_exhaustive()
+    }
+}
+
+impl<T> Deref for MMapRegion<T> {
+    type Target = [T];
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<T> Default for Region<T> {
+    #[inline]
+    fn default() -> Self {
+        Self::new_empty()
+    }
+}
+
+impl<T> Region<T> {
+    /// Create a new empty region.
+    #[inline]
+    #[must_use]
+    pub fn new_empty() -> Region<T> {
+        Region::Heap(Vec::new())
+    }
+
+    /// Create a new heap-allocated region of a specific capacity.
+    #[inline]
+    #[must_use]
+    pub fn new_heap(capacity: usize) -> Region<T> {
+        Region::Heap(Vec::with_capacity(capacity))
+    }
+
+    /// Create a new file-based mapped region of a specific capacity. The capacity of the
+    /// returned region can be larger than requested to accommodate page sizes.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the memory allocation fails.
+    #[inline(always)]
+    pub fn new_mmap(capacity: usize) -> Result<Region<T>, lgalloc::AllocError> {
+        lgalloc::allocate(capacity).map(|(ptr, capacity, handle)| {
+            // SAFETY: memory points to suitable memory.
+            // It is UB to call `from_raw_parts` with a pointer not allocated from the system
+            // allocator, but we accept this here because we promise never to reallocate the vector.
```

Comment:
I was confused by the system allocator requirement, because I understand that to mean the allocator used by the OS. But I assume you mean the global allocator?

Reply:
Thanks, fixed!